signals/signal_exporter.py

```diff
@@ -0,0 +1,421 @@
+        hostname to be inputted manually (otherwise will default to 
+        platform.node() value)
+        """
+        self.subs = []
+        self.proc_name = process_name
+        self.runner_host = runner_host
+        self.pub_id = process_name + "-" + str(uuid.uuid4())
+        self.redis = redis.Redis(host=redis_host, port=redis_port, db=0)
+        self.init_listener = None
+        self.legal_events = None
+
+    def _sig_builder(
+        self, event: str, sample: int = -1, tag: str = None, metadata: Dict = None
+    ) -> Signal:
+        """
+        Build a signal data object based on exporter object fields,
+        as well as user-inputted fields. Returns the signal object.
+        """
+        config = {
+            "publisher_id": self.pub_id,
+            "process_name": self.proc_name,
+            "event": event,
+            "runner_host": self.runner_host,
+        }
+        if sample:
+            config["sample_no"] = sample
+        if tag:
+            config["tag"] = tag
+        if metadata:
+            config["metadata"] = metadata
+        sig = Signal(**config)
+        return sig
+
+    def _get_data_dict(self, response: Dict) -> Dict:
+        """
+        Returns response signal payload if a properly-formed
+        response is received. Otherwise return None.
+        """
+        if not "data" in response:
+            print("No data in this response message")
+            return None
+        try:
+            data = json.loads(response["data"])
+        except ValueError:
+            return None
+        if (
+            "responder_id" not in data
+            or "publisher_id" not in data
+            or "event" not in data
+        ):
+            print("Malformed response data found")
+            return None
+        return data
+
+    def _fetch_responders(self) -> None:
+        """
+        Start initialization response listener. Add respoder_ids from proper
+        responses to the subscriber list.
+        """
+        subscriber = self.redis.pubsub(ignore_subscribe_messages=True)
+
+        def _init_handler(item) -> None:
+            data = self._get_data_dict(item)
+            if (
+                data
+                and data["event"] == "initialization"
+                and data["publisher_id"] == self.pub_id
+            ):
+                self.subs.append(data["responder_id"])
+
+        subscriber.subscribe(**{"event-signal-response": _init_handler})
+        self.init_listener = subscriber.run_in_thread()
+
+    def _check_subs(self, event: str) -> Tuple[Any, List[int]]:
+        """
+        Listen for responses from all registered subscribers. Return
+        listener, as well as value based on responders' RAS codes.
+        """
+        if not self.subs:
+            return None, [0]
+
+        to_check = set(self.subs)
+        subscriber = self.redis.pubsub(ignore_subscribe_messages=True)
+        result_code_holder = [ResultCodes.ALL_SUBS_SUCCESS]
```

Comment:
Love what you did here, adding the ``_holder`` makes this easier to read through and understand what is going on. I did find a snippet from Google's Python style guide that warns on relying on the atomic operations of built-ins, however I think for this use case it should be all good: https://google.github.io/styleguide/pyguide.html#218-threading